### PR TITLE
Trn 396 dinov2 allow vit model training

### PR DESCRIPTION
**Pull request by JonasWurst**
## What has changed and why?
This PR brings a new model package that wraps the dinov2_vit.

## How has it been tested?
Unit tests are in place

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [x] Not needed (internal change)

## Did you update the documentation?

- [ ] Yes
- [x] Not needed (internal change without effects for user)

### CODE REVIEW

**Comment by JonasWurst** [comment]
NOTE: exporting and testing of it, not yet done.

File: src/lightly_train/_models/dinov2_vit/dinov2_vit_package.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        return DINOv2ViTModelWrapper(model)
+
+    @classmethod
+    def export_model(cls, model: Module, out: Path, log_example: bool = True) -> None:
+        torch.save(model.state_dict(), out)
+        #TODO: finish implement export_model
+
+# Create singleton instance of the package. The singleton should be used whenever
+# possible.
+DINOV2_VIT_Package_PACKAGE = DINOv2ViTPackage()
```
Comment (guarin) [inline src/lightly_train/_models/dinov2_vit/dinov2_vit_package.py]:
```suggestion
DINOV2_VIT_PACKAGE = DINOv2ViTPackage()
```

File: src/lightly_train/_models/dinov2_vit/dinov2_vit.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    def forward_pool(self, x: ForwardFeaturesOutput) -> ForwardPoolOutput:
+        return {"pooled_features": self._pool(x["features"])}
+
+    def make_teacher(self) -> None:
+        if self._model.chunked_blocks:
+            for chunked_blocks in self._model.blocks:
+                update_blocks_student_to_teacher(chunked_blocks)
+        else:
+            update_blocks_student_to_teacher(self._model.blocks)             
+
```
Comment (guarin) [inline src/lightly_train/_models/dinov2_vit/dinov2_vit.py]: Please add `get_model` function as in https://github.com/lightly-ai/lightly-train/pull/52

**Review by guarin — CHANGES_REQUESTED** [review]

File: tests/_models/dinov2_vit/test_dinov2_vit.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+from lightly_train._modules.teachers.dinov2.layers.drop_path import DropPath
+from lightly_train._modules.teachers.dinov2.models.vision_transformer import (
+    DinoVisionTransformer,
+)
+from lightly_train._modules.teachers.dinov2.models.vision_transformer import (
+    vit_small as vit_small_untyped,
+)
+
+
+def vit_small() -> DinoVisionTransformer:
```
Comment (yutong-xiang-97) [inline tests/_models/dinov2_vit/test_dinov2_vit.py]: Should we alternatively add typing to the original functions?

File: src/lightly_train/_models/dinov2_vit/dinov2_vit.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            raise ValueError(
+                f"Unexpected shape for x_norm_patchtokens: {rt['x_norm_patchtokens'].shape}"
+            )
+        return {"features": features_reshaped, "cls_token": rt["x_norm_clstoken"]}
+    
+    def forward_pool(self, x: ForwardFeaturesOutput) -> ForwardPoolOutput:
+        return {"pooled_features": self._pool(x["features"])}
+    
+    def get_model(self) -> DinoVisionTransformer:
+        return cast(DinoVisionTransformer, self._model)
```
Comment (yutong-xiang-97) [inline src/lightly_train/_models/dinov2_vit/dinov2_vit.py]: Personally, I would prefer
```
model: DinoVisionTransformer = self._model
return model
```
instead of casting to be more type safe

File: src/lightly_train/_models/dinov2_vit/dinov2_vit.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        super().__init__()
+        self._model = model
+        self._feature_dim = int(self._model.embed_dim)
+        self._pool = AdaptiveAvgPool2d((1, 1))
+    
+    def feature_dim(self) -> int:
+        return self._feature_dim
+
+    def forward_features(self, x: Tensor) -> ForwardFeaturesOutput:
+        rt = self._model(x, is_training=True) # forcing to return all patches
```
Comment (yutong-xiang-97) [inline src/lightly_train/_models/dinov2_vit/dinov2_vit.py]: Personally I would prefer directly using
```
self._model.forward_features()
```
https://github.com/lightly-ai/lightly-train/blob/35f3004f495430b2056eb1ecc4fa11052ea6547c/src/lightly_train/_modules/teachers/dinov2/models/vision_transformer.py#L308C9-L308C25

File: src/lightly_train/_models/dinov2_vit/dinov2_vit.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        if rt["x_norm_patchtokens"].dim() == 3:
+            patches_resolution = self._model.patch_embed.patches_resolution
+            features_reshaped = rt["x_norm_patchtokens"].reshape(
+                rt["x_norm_patchtokens"].shape[0],
+                rt["x_norm_patchtokens"].shape[2],
+                patches_resolution[0],
+                patches_resolution[1],
+            )
+        elif rt["x_norm_patchtokens"].dim() == 4:
```
Comment (yutong-xiang-97) [inline src/lightly_train/_models/dinov2_vit/dinov2_vit.py:41]: Just curious: in which case would the dim be 4?

File: src/lightly_train/_models/dinov2_vit/dinov2_vit_package.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            "vit_giant2": vits.vit_giant2,
+        }
+        model_builder = model_builders.get(cfg.student.arch, None)
+        if model_builder is None:
+            raise TypeError(
+                f"Unsupported architecture type {cfg.student.arch}."
+            )
+        
+        # Cast the model builder to the correct type
+        model_builder = cast(
```
Comment (yutong-xiang-97) [inline src/lightly_train/_models/dinov2_vit/dinov2_vit_package.py]: Same question as before, should we cast it dynamically or we use static type checking for the model_builder functions?

File: src/lightly_train/_models/dinov2_vit/dinov2_vit_package.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            Callable[...,DinoVisionTransformer],
+            model_builder
+        )
+        
+        model = model_builder(
+            img_size=cfg.crops.global_crops_size,
```
Comment (yutong-xiang-97) [inline src/lightly_train/_models/dinov2_vit/dinov2_vit_package.py:70]: The `DinoVisionTransformer` has more init params than listed here. Is it deliberate that we fixed the ones that are configurable and use the defaults for the rest? If not, I think using `**kwargs` makes sense here instead of using exact match of kwargs.

File: src/lightly_train/_modules/teachers/dinov2/configs/__init__.py [hunk 9]
```diff
@@ -9,6 +9,39 @@
 
 from omegaconf import OmegaConf
 
+MODELS = {
```
Comment (yutong-xiang-97) [inline src/lightly_train/_modules/teachers/dinov2/configs/__init__.py:16]: I don't see serious issues with putting the config loading logic in `__init__.py` for now, esp. when it is fixed, but `__init__.py` was not supposed to handle code as it might have some implications.

File: src/lightly_train/_models/dinov2_vit/dinov2_vit.py (hunk 4 above)
Comment (JonasWurst) [reply src/lightly_train/_models/dinov2_vit/dinov2_vit.py]: Cast is simply telling for unytped vars, what to expect. So in some place, we have to perform the cast, if we are not changing the typing in the base implementation. So far we decided to not dive into that rabbit hole.

File: src/lightly_train/_models/dinov2_vit/dinov2_vit.py (hunk 6 above)
Comment (JonasWurst) [reply src/lightly_train/_models/dinov2_vit/dinov2_vit.py:41]: If already reshaped

File: tests/_models/dinov2_vit/test_dinov2_vit.py (hunk 3 above)
Comment (JonasWurst) [reply tests/_models/dinov2_vit/test_dinov2_vit.py]: Initially I refrained doin it, since we also did not touch the vit code in the previous PRs. However, I decided it is simpler to break that rule, so the remaining code becomes simpler.

File: src/lightly_train/_models/dinov2_vit/dinov2_vit.py (hunk 4 above)
Comment (JonasWurst) [reply src/lightly_train/_models/dinov2_vit/dinov2_vit.py]: See comment above, I've changed the type in vit code, hence cast is not required anymore

File: src/lightly_train/_models/dinov2_vit/dinov2_vit.py (hunk 5 above)
Comment (JonasWurst) [reply src/lightly_train/_models/dinov2_vit/dinov2_vit.py]: I personally see no benefit in doing it. The typical thing which is used is the forward. I would keep it as is

File: src/lightly_train/_models/dinov2_vit/dinov2_vit_package.py (hunk 7 above)
Comment (JonasWurst) [reply src/lightly_train/_models/dinov2_vit/dinov2_vit_package.py]: resolved, since took the liberty to change the vit code

File: src/lightly_train/_models/dinov2_vit/dinov2_vit_package.py (hunk 8 above)
Comment (JonasWurst) [reply src/lightly_train/_models/dinov2_vit/dinov2_vit_package.py:70]: I'm now combining the two dicts, however I think it is not a good idea to play around with that and overwrite the cfg.

File: tests/_models/dinov2_vit/test_dinov2_vit.py (hunk 3 above)
Comment (guarin) [reply tests/_models/dinov2_vit/test_dinov2_vit.py]: > Initially I refrained doin it, since we also did not touch the vit code in the previous PRs. However, I decided it is simpler to break that rule, so the remaining code becomes simpler.

Please add disclaimers in the code in case you modify any code that is not from us

File: src/lightly_train/_modules/teachers/dinov2/configs/__init__.py (hunk 9 above)
Comment (JonasWurst) [reply src/lightly_train/_modules/teachers/dinov2/configs/__init__.py:16]: This is again down to not refactoring the vit codebase much. They already had logic related to cfg loading. Hence that was the best suited place w/o messing with the codebase.

But in general, I agree we should keep it clean.

File: tests/_models/dinov2_vit/test_dinov2_vit.py (hunk 3 above)
Comment (JonasWurst) [reply tests/_models/dinov2_vit/test_dinov2_vit.py]: Last time I've changed things, I'v mentioned it in the Notice file. Is that sufficient in your opinion?

Comment (guarin) [reply tests/_models/dinov2_vit/test_dinov2_vit.py]: If it is bigger things (add/remove/modify function) I would add a note in the code.

File: src/lightly_train/_models/dinov2_vit/dinov2_vit.py (hunk 6 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_models/dinov2_vit/dinov2_vit.py:41]: In which case is it necessary to reshape the features? In original DINOv2 it is used directedly: https://github.com/facebookresearch/dinov2/blob/e1277af2ba9496fbadf7aec6eba56e8d882d1e35/dinov2/train/ssl_meta_arch.py#L165

File: src/lightly_train/_models/dinov2_vit/dinov2_vit_package.py (hunk 8 above)
Comment (JonasWurst) [reply src/lightly_train/_models/dinov2_vit/dinov2_vit_package.py:70]: Remark: the kwargs matching is copied from the original codebase

**Review by yutong-xiang-97 — APPROVED** [review]
